stan/math/opencl/copy.hpp

```diff
@@ -36,22 +39,24 @@ void copy(matrix_cl& dst, const Eigen::Matrix<double, R, C>& src) {
                    "dst.rows()", dst.rows());
   check_size_match("copy (Eigen -> (OpenCL))", "src.cols()", src.cols(),
                    "dst.cols()", dst.cols());
-  if (src.size() == 0) {
-    return;
-  }
-  cl::CommandQueue queue = opencl_context.queue();
-  try {
-    /**
-     * Writes the contents of src to the OpenCL buffer
-     * starting at the offset 0
-     * CL_TRUE denotes that the call is blocking
-     * We do not want to execute any further kernels
-     * on the device until we are sure that the data is transferred)
-     */
-    queue.enqueueWriteBuffer(dst.buffer(), CL_TRUE, 0,
-                             sizeof(double) * dst.size(), src.data());
-  } catch (const cl::Error& e) {
-    check_opencl_error("copy Eigen->(OpenCL)", e);
+  if (src.size() > 0) {
+    try {
+      /**
+       * Writes the contents of src to the OpenCL buffer
+       * starting at the offset 0
+       * CL_FALSE denotes that the call is non-blocking
+       * This means that future kernels need to know about the copy_event
+       * So that they do not execute until this transfer finishes.
+       */
+      cl::CommandQueue queue = opencl_context.queue();
+      cl::Event copy_event;
+      queue.enqueueWriteBuffer(dst.buffer(), CL_FALSE, 0,
+                               sizeof(double) * dst.size(), src.data(), NULL,
+                               &copy_event);
+      dst.add_event<event_cl::write>(copy_event);
```

Comment:
Yeah that makes sense, Bob gave me the C++ templates book so was in template mindspace

Reply:
Now that I'm not at work I had enough time to have this make less sense.

> In general we don't want to use templates unless there's no other way (for a given performance goal or whatever). 

Why? tbh I like how it looks. But also I originally had add_write and add_read functions, but I kept getting errors when trying to do the template specializations for `assign_event` for `write_buffer` and `read_buffer`

>I also don't like having a default event type - better to just have two methods and always be explicit.

100% agree with this

Reply:
We want to avoid templates unless they're necessary because they're the strongest and least safe parts of the language. Everything about working with templates is harder or worse - design, reasoning through behavior, debugging, compilation times get much longer, etc.

re: 2 vs. 1: you are defining two methods, just using templates instead of writing them out. Sometimes this is what you want, but that's mostly when you want your method to work with a variety of types that don't have a shared interface. We sometimes use templates to switch at compile time to provide different methods for different types, but I don't think we ever use templates to do different things for the same type. We're going to know at each use-site whether we want to read or write, and we're even going to write it down - there's no reason to use a template for that. Stick with the least powerful subset of the language that does the job reasonably well.

Reply:
That's fair I agree with this

Reply:
Wrote out a bunch of getters and setters for the read and read_write event queues. In terms of templating, code has gone from SSJ Goku to a bunch of Saibamen